arch/xtensa/core/README-WINDOWS.rst

```diff
@@ -0,0 +1,108 @@
+in two places.  They have to be stored on the stack across potentially
+many calls, so they need to be GPR data that lives in registers and
+can be spilled.  But ENTRY isn't specified to assume a particular
+return value format and is used immediately, so it makes more sense
+for it to use processor state instead.
+
+Note that we still don't know how to detect when the register file has
+wrapped around and needs to be spilled or filled.  To do this there is
+a WINDOWSTART register used to detect which register quads are in use.
+The name "start" is somewhat confusing, this is not a pointer.
+WINDOWSTART stores a bitmask with one bit per hardware quad (so it's 8
+or 16 bits wide).  The bit in windowstart corresponding to WINDOWBASE
+will be set by the ENTRY instruction, and remain set after rotations
+until cleared by a function return (by RETW, see below).  Other bits
+stay zero.  So there is one set bit in WINDOWSTART corresponding to
+each call frame that is live in hardware registers, and it will be
+followed by 0, 1 or 2 zero bits that tell you how "big" (how many
+quads of registers) that frame is.
+
+So the CPU executing RETW checks to make sure that the register quad
+being brought into A0-A3 (i.e. the new WINDOWBASE) has a set bit
+indicating it's valid. If it does not, the registers must have been
+spilled and the CPU traps to an exception handler to fill them.
+
+Likewise, the processor can tell if a high register is "owned" by
+another call by seeing if there is a one in WINDOWSTART between that
+register's quad and WINDOWBASE.  If there is, the CPU traps to a spill
+handler to spill one frame.  Note that a frame might be only four
+registers, but it's possible to hit registers 12 out from WINDOWBASE,
+so it's actually possible to trap again when the instruction restarts
+to spill a second quad, and even a third time at maximum.
+
+Finally: note that hardware checks the two bits of WINDOWSTART after
+the frame bit to detect how many quads are represented by the one
+frame.  So there are six separate exception handlers to spill/fill
+1/2/3 quads of registers.
+
+## Software & ABI
+
+The advantage of the scheme above is that it allows the registers to
+be spilled naturally into the stack by using the stack pointers
+embedded in the register file.  But the hardware design assumes and to
+some extent enforces a fairly complicated stack layout to make that
+work:
+
+The spill area for a single frame's A0-A3 registers is not in its own
+stack frame.  It lies in the 16 bytes below its CALLEE's stack
+pointer.  This is so that the callee (and exception handlers invoked
+on its behalf) can see its caller's potentially-spilled stack pointer
+register (A1) on the stack and be able to walk back up on return.
+Other architectures do this too by e.g. pushing the incoming stack
+pointer onto the stack as a standard "frame pointer" defined in the
+platform ABI.  Xtensa wraps this together with the natural spill area
+for register windows.
```

Comment:
@andyross fwiw, I've noticed some differences between GCC and XCC here wrt stack and register layout when debugging exceptions. The SOF exception code will dump registers and stack but this looks different between XCC and GCC. SOF has a python script to decode the dump (works on XCC not GCC). Something to be aware off here for later on.